beets/dbcore/types.py

```diff
@@ -37,8 +43,12 @@ def __init__(self, value: Any = None): ...
 # Generic type variables, used for the value type T and null type N (if
 # nullable, else T and N are set to the same type for the concrete subclasses
 # of Type).
-N = TypeVar("N")
-T = TypeVar("T", bound=ModelType)
+if TYPE_CHECKING:
+    N = TypeVar("N", default=Any)
+    T = TypeVar("T", bound=ModelType, default=Any)
+else:
+    N = TypeVar("N")
+    T = TypeVar("T", bound=ModelType)
```

Comment:
See above regarding `default` (here, in particular, I'd argue that there should be no default, such that the type checker forces one to specify these types).

Reply:
As I mentioned this has to do with those type annotations where we use the generic type, in this case `Type`. For example, let's take `db.py` which has these annotations:

```py
    _fields: dict[str, types.Type] = {}
    _types: dict[str, types.Type] = {}
    def _type(cls, key) -> types.Type:
    def _make_table(self, table: str, fields: Mapping[str, types.Type]):
```

By default `mypy` is unhappy with `Missing type parameters for generic type "Type"`

A couple of possible ways to deal with this:

1. Explicit `Any`
```py
    _fields: dict[str, types.Type[Any]] = {}
    _types: dict[str, types.Type[Any]] = {}
    def _type(cls, key) -> types.Type[Any]:
    def _make_table(self, table: str, fields: Mapping[str, types.Type[Any]]):
```

2. Explicit _dry_ `Any`
```py
AnyType = types.Type[Any]

    _fields: dict[str, AnyType] = {}
    _types: dict[str, AnyType] = {}
    def _type(cls, key) -> AnyType:
    def _make_table(self, table: str, fields: Mapping[str, AnyType]):
```

3. `Any` defined as default on the typevar

```py
T = TypeVar("T", bound=ModelType, default=Any)
```

4. Or just disable this check with `allow_any_generics = true`


The main issue with **1** is that it adds complexity to the type definitions and creates noise. **2** is slightly less bad, however we'd need to create such variable for each generic type we want to use, it needs importing and adds indirecttion.

I found that specifying the default on the typevar addresses most of the issues above - we can use the actual type directly, it doesn't add any noise and we don't need to define additional variables. We do need to define it in a funny way using `TYPE_CHECKING` however 😅 

Of course we could always just disable this check, but this will affect all generic types, and mypy won't anymore enforce specific types for `list`, `dict` etc.

Reply:
Yes, let's stick to 3, cf. my comment in another thread.